### ISSUE TITLE
Revisit metadata

### ISSUE THREAD

**Issue by marcofucci** [uktrade/data-hub-leeloo#358]
## The problem
At the moment:
1. metadata is a separate but centralised app containing core metadata (e.g. `Team`, `Sector`, `Country`) and app specific metadata (e.g. `InvestmentType`, `InvestmentStrategicDriver`)
1. the metadata _API_ endpoints are centralised
1. `core.constants` is a centralised place for Enums of _metadata_ models
1. `fixtures` is centralised and contains all metadata fixtures

The problem is that this does not encapsulate well the specific apps (_company_, _investment_, _omis_).

E.g. let's say that we need to add a new type of metadata called `OrderCategory` for the omis app.

In order to do that, we need to create a new model in `metadata.models`, configure the `metadata` API to include the new type, add the new type to `constants` and add extra `fixtures`.

However, this new model only makes sense in the context of the omis app and this breaks encapsulation.

## Proposed solution (to be discussed)

1. Metadata models are app specific and not centralised.
1. the API endpoints are not centralised but app specific
1. `constants` is not centralised but app specific
1. `fixtures` are not centralised (this could be potentially if we manage things with sub folders)

The API could look like:

```
/v3/omis/order
/v3/omis/order/<pk>
/v3/omis/order/<order_pk>/subscriber-list
/v3/omis/order/metadata/order-category

/v3/investment
/v3/investment/<pk>
/v3/investment/<pk>/archive
/v3/investment/<pk>/audit
/v3/investment/<pk>/unarchive
/v3/investment/<project_pk>/document
/v3/investment/<project_pk>/document/<doc_pk>
/v3/investment/<project_pk>/document/<doc_pk>/upload-callback
/v3/investment/<project_pk>/team-member
/v3/investment/<project_pk>/team-member/<adviser_pk>
/v3/investment/metadata/investment-project-stage
/v3/investment/metadata/investment-type
/v3/investment/metadata/fdi-type
/v3/investment/metadata/non-fdi-type
/v3/investment/metadata/investment-business-activity
/v3/investment/metadata/investment-strategic-driver
/v3/investment/metadata/fdi-value
/v3/investment/metadata/salary-range

/v3/ch-company
/v3/ch-company/<company_number>
/v3/company
/v3/company/<pk>
/v3/company/<pk>/archive
/v3/company/<pk>/audit
/v3/company/<pk>/unarchive
/v3/company/metadata/company-classification
/v3/company/metadata/headquarter-type
/v3/company/metadata/business-type
/v3/company/metadata/employee-range
/v3/company/metadata/turnover

/v3/contact
/v3/contact/<pk>
/v3/contact/<pk>/archive
/v3/contact/<pk>/audit
/v3/contact/<pk>/unarchive
/v3/contact/metadata/role (we might want to rename this to contact-role)

/v3/adviser
/v3/adviser/<pk>
/v3/adviser/metadata/team
/v3/adviser/metadata/team-role

/v3/interaction
/v3/interaction/<pk>
/v3/interaction/metadata/interaction-type

(or /v3/core/metadata/...)
/v3/metadata/country
/v3/metadata/title
/v3/metadata/uk-region
/v3/metadata/sector
/v3/metadata/referral-source-marketing (might go in investment)
/v3/metadata/referral-source-activity (might go in investment)
/v3/metadata/referral-source-website (might go in investment)

/v3/service-delivery
/v3/service-delivery/<pk>
/v3/service/metadata/service
/v3/service/metadata/service-delivery-status
/v3/service/metadata/event (we might want to rename this)

/v3/business-leads
/v3/business-leads/<pk>
/v3/business-leads/<pk>/archive
/v3/business-leads/<pk>/unarchive

/v3/search
/v3/search/company
/v3/search/contact
/v3/search/investment_project
```

**Comment by jim68000** [comment]
Seems sensible. I'd agree that the referral source metadata belongs in investment (there are other referral sources, but they're banks, just to make life more complicated). Need to keep a keen eye on whether there are other pieces of metadata shared across apps, and whether it's sensible to centralise them - I think you have the right ones here, but I'll have a deeper thing.

Also inclined to think that some of these things shouldn't be metadata anyway - teams, for example, need more modelling (a team should have a location as well as a name, at the very least). And we've discussed whether sector needs to be hierarchical elsewhere.

**Comment by marcofucci** [comment]
OK I moved the referral source to /investment/metadata, thanks @jim68000

**Comment by marcofucci** [comment]
and yes, I agree that `team` could be a separate app

**Comment by reupen** [comment]
I also don't like the duplication of the metadata in the YAML files and in the constants – it's quite painful maintenance-wise.

We might be able to get rid of Role (= contact role) as it doesn't seem to be used as a foreign key anywhere. Title is also no longer in the front end, so might be able to prune it.

I also question if we have two types of metadata:
- Things that an admin might want to edit (e.g. teams)
- Things that shouldn't really be touched via the admin pages (e.g. investment type, investment project stage)

**Comment by marcofucci** [comment]
@reupen 
1. `Role` seems to be implicitly used in `Contact.job_title`. We might have to fix the declaration of `Contact` to make that obvious?

2. `Title` seems to be used in the frontend when dealing with contacts. However if we manage to agree on removing it, it'd make things cleaner

**Comment by marcofucci** [comment]
I agree that we have 2 types of metadata and the 2nd group doesn't really have to be in tables?

**Comment by elcct** [comment]
If the 2nd group is not going to be maintained via admin pages then it makes sense to leave it out of the database. 
@marcofucci proposal makes sense to me and it should make metadata easier to work with.
I also think teams don't belong in metadata.

**Comment by marcofucci** [comment]
I've added an issue #359 related to `Team` and `TeamRole` as we all agree on that and we can keep it out from this main issue.

I'm still not 100% convinced by this way, let's keep thinking about it.

**Comment by reupen** [comment]
> Role seems to be implicitly used in Contact.job_title. We might have to fix the declaration of Contact to make that obvious?

Some tests are using the constants, but job title is a text field in the front end.

> Title seems to be used in the frontend when dealing with contacts. However if we manage to agree on removing it, it'd make things cleaner

There might be some logic left over there, but it's not displayed or part of the form in the UI (it was removed). There's a chance it might be added again if turns out it's needed, but not without some reviewing of it (the list of titles was too long).

**Comment by marcofucci** [comment]
@reupen 💯 
Re. `Role`, I've created another issue #360
Re. `Title`, I've created another issue #361

**Comment by marcofucci** [comment]
Ops, sorry for the noise, we were testing the github API when assigning people and didn't realise that I was polluting the issue 😬

**Comment by marcofucci** [comment]
I thought about it and I think this is trying to solve too many problems at the same time.

I think we could keep the metadata API as they are and fix the following issues:

1. metadata models should be moved from the `metadata` app to their specific app. 
E.g. 
`metadata.models.CompanyClassification` will be moved to `company.models.CompanyClassification`

2. the related constants should be moved from `core.constants` to their specific app. 
Later on we might decide how we want to fix this dual existence as I don't particularly like it as well @reupen 

This means that for now we would:

- keep the API centralised in `metadata/<...>` and maybe have a way of registering metadata automatically so that they don't have to be hardcoded in the `metadata` django app.

- maybe keep the `fixtures` centralised.

And discuss these points at a later stage.

What do you guys think?

**Comment by elcct** [comment]
That sounds good to me.
